### PR TITLE
LWG Poll 6: P1963R0 Fixing US 313

### LINKED ISSUE #3708: [2020-02 LWG Poll 6] P1963R0 Fixing US 313
resolves US313 and LWG issue 3156

https://wg21.link/P1963R0

### PR DESCRIPTION
**Pull request by jensmaurer**
Also fixes NB US 313 (C++20 CD) and LWG3156.

Fixes #3708.
Fixes cplusplus/papers#697.
Fixes cplusplus/nbballot#309.

### CODE REVIEW

File: source/utilities.tex [hunk 1]
```diff
@@ -6746,16 +6760,16 @@
...
-    constexpr ForwardIterator destroy_n(ForwardIterator first, Size n);
-  template<class ExecutionPolicy, class ForwardIterator, class Size>
-    constexpr ForwardIterator destroy_n(ExecutionPolicy&& exec, // see \ref{algorithms.parallel.overloads}
-                                        ForwardIterator first, Size n);
+                           NoThrowForwardIterator first, NoThrowForwardIterator last);
+  template<class NoThrowForwardIterator, class Size>
+    constexpr NoThrowForwardIterator destroy_n(NoThrowForwardIterator first, Size n);
+  template<class ExecutionPolicy, class NoThrowForwardIterator, class Size>
+    constexpr NoThrowForwardIterator destroy_n(ExecutionPolicy&& exec,  // see \ref{algorithms.parallel.overloads}
+                                        NoThrowForwardIterator first, Size n);
```
Comment (burblebee) [inline source/utilities.tex]: indent

File: source/algorithms.tex [hunk 2]
```diff
@@ -88,6 +88,14 @@
   \tcode{Forward\-Iterator2},
   the template argument shall meet the
   \oldconcept{ForwardIterator} requirements\iref{forward.iterators}.
+\item
+  If an algorithm’s template parameter is named
```
Comment (burblebee) [inline source/algorithms.tex]: Unicode quote used?  rendered as "algorithmâĂŹs"

File: source/algorithms.tex [hunk 3]
```diff
@@ -88,6 +88,14 @@
...
+  \tcode{NoThrowForwardIterator},
+  the template argument shall meet the
+  \oldconcept{ForwardIterator} requirements\iref{forward.iterators}, and
+  is required to have the property that no exceptions are thrown
+  from increment, assignment, comparison, or
+  indirection through valid iterators.
```
Comment (burblebee) [inline source/algorithms.tex]: The wording here doesn't make sense, can it be fixed?

File: source/algorithms.tex [hunk 4]
```diff
@@ -9917,6 +9925,714 @@
...
+\end{codeblock}
+\end{itemdescr}
+
+\rSec2[uninitialized.copy]{\tcode{uninitialized_copy}}
+
+\indexlibraryglobal{uninitialized_copy}%
+\begin{itemdecl}
+template<class InputIterator, class NoThrowForwardIterator>
+  NoThrowForwardIterator uninitialized_copy(InputIterator first, InputIterator last,
+                                     NoThrowForwardIterator result);
```
Comment (burblebee) [inline source/algorithms.tex]: indent

File: source/algorithms.tex [hunk 5]
```diff
@@ -9917,6 +9925,714 @@
...
+\end{codeblock}
+\end{itemdescr}
+
+\rSec2[uninitialized.move]{\tcode{uninitialized_move}}
+
+\indexlibraryglobal{uninitialized_move}%
+\begin{itemdecl}
+template<class InputIterator, class NoThrowForwardIterator>
+  NoThrowForwardIterator uninitialized_move(InputIterator first, InputIterator last,
+                                     NoThrowForwardIterator result);
```
Comment (burblebee) [inline source/algorithms.tex]: indent

File: source/algorithms.tex (hunk 2 above)
Comment (johelegp) [reply source/algorithms.tex]: Yes. The automatic check could be enhanced to detect this.

File: source/algorithms.tex (hunk 3 above)
Comment (jensmaurer) [reply source/algorithms.tex]: My guess is you're reading this as "comparison through valid iterators".

File: source/algorithms.tex (hunk 4 above)
Comment (jensmaurer) [reply source/algorithms.tex]: Fixed.

File: source/utilities.tex (hunk 1 above)
Comment (jensmaurer) [reply source/utilities.tex]: Fixed.

File: source/algorithms.tex (hunk 3 above)
Comment (jensmaurer) [reply source/algorithms.tex]: Fixed in separate commit so @zygoloid can omit it if it's too much editorial fixup.

File: source/algorithms.tex [hunk 6]
```diff
@@ -9767,7 +9775,7 @@
 \tcode{result} may be equal to \tcode{first}.
 For the overloads with an \tcode{ExecutionPolicy},
 the ranges \range{first}{last} and \range{result}{result + (last - first)}
-shall not overlap.
+does not overlap.
```
Comment (CaseyCarter) [inline source/algorithms.tex]: This is normative damage: this paragraph imposes a requirement on values provided by the user of the library, so it should use "shall".

File: source/algorithms.tex [hunk 7]
```diff
@@ -9917,6 +9925,714 @@
...
+\begin{itemdecl}
+template<class InputIterator, class Size, class NoThrowForwardIterator>
+  NoThrowForwardIterator uninitialized_copy_n(InputIterator first, Size n,
+                                              NoThrowForwardIterator result);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\range{result}{n} does not overlap with \range{first}{n}.
```
Comment (CaseyCarter) [inline source/algorithms.tex]: Per #2932:
```suggestion
\countedrange{result}{n} does not overlap with \countedrange{first}{n}.
```

File: source/algorithms.tex [hunk 8]
```diff
@@ -9917,6 +9925,714 @@
...
+    uninitialized_copy_n_result<I, O>
+      uninitialized_copy_n(I ifirst, iter_difference_t<I> n, O ofirst, S olast);
+}
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\range{ofirst}{olast} does not overlap with
+\range{ifirst}{n}.
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
\countedrange{ifirst}{n}.
```

File: source/algorithms.tex [hunk 9]
```diff
@@ -9917,6 +9925,714 @@
...
+\begin{itemdecl}
+template<class InputIterator, class NoThrowForwardIterator>
+  NoThrowForwardIterator uninitialized_move(InputIterator first, InputIterator last,
+                                            NoThrowForwardIterator result);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\range{result}{(last - first)} does not overlap with \range{first}{last}.
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
\countedrange{result}{(last - first)} does not overlap with \range{first}{last}.
```

File: source/algorithms.tex [hunk 10]
```diff
@@ -9917,6 +9925,714 @@
...
+\begin{itemdecl}
+template<class InputIterator, class Size, class NoThrowForwardIterator>
+  pair<InputIterator, NoThrowForwardIterator>
+    uninitialized_move_n(InputIterator first, Size n, NoThrowForwardIterator result);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\range{result}{n} does not overlap with \range{first}{n}.
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
\countedrange{result}{n} does not overlap with \countedrange{first}{n}.
```

File: source/algorithms.tex [hunk 11]
```diff
@@ -9917,6 +9925,714 @@
...
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+for (; n > 0; ++result, (void) ++first, --n)
+  ::new (@\placeholdernc{voidify}@(*result))
+    typename iterator_traits<NoThrowForwardIterator>::value_type(std::move(*first));
+return {first,result};
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
return {first, result};
```

File: source/algorithms.tex [hunk 12]
```diff
@@ -9917,6 +9925,714 @@
...
+    uninitialized_move_n_result<I, O>
+      uninitialized_move_n(I ifirst, iter_difference_t<I> n, O ofirst, S olast);
+}
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\range{ofirst}{olast} does not overlap with
+\range{ifirst}{n}.
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
\countedrange{ifirst}{n}.
```

File: source/algorithms.tex [hunk 13]
```diff
@@ -9917,6 +9925,714 @@
...
+\begin{codeblock}
+auto t = uninitialized_move(counted_iterator(ifirst, n),
+                            default_sentinel, ofirst, olast);
+return {t.in.base(), t.out};
+\end{codeblock}
+
+\pnum
+\begin{note}
+If an exception is thrown, some objects in the range
+\range{first}{n}
```
Comment (CaseyCarter) [inline source/algorithms.tex]:
```suggestion
\countedrange{first}{n}
```

File: source/algorithms.tex (hunk 6 above)
Comment (jwakely) [reply source/algorithms.tex]: And "the ranges A and B does not overlap" is incorrect grammar anyway.

**Comment by jensmaurer** [comment]
@zygoloid , if the two fixup patches are too big to include right here, we should move them to a separate fixup pull request.

**Pull request merged**